File: day4.py
import aoc
from itertools import product


def is_in_bounds(lines, x, y):
    return 1 <= y < len(lines) and 0 <= x < len(lines[y])


def is_word(lines, word, x, y, direction):
    if not word:
        return True
    elif is_in_bounds(lines, x, y) and lines[y][x] == word[0]:
        x = x + direction[0]
        y += direction[1]
        return is_word(lines, word[1:], x, y, direction)
    else:
        return False


def solve(lines: list[str]) -> int:
    sum = 0
    for y in range(len(lines)):
        for x in range(len(lines[y])):
            for direction in product((-1, 0, 1), (-1, 0, 1)):
                sum += is_word(lines, 'XMAS', x, y, direction)
    return sum


def solve2(lines: list[str]) -> int:
    sum = 0
    for y in range(len(lines)):
        for x in range(len(lines[y])):
            crosses = 0
            for direction in ((-1, -1), (1, -1), (1, 1), (-1, 1)):
                start_x = x - direction[0]
                start_y = y - direction[1]
                crosses += is_word(lines, 'MAS', start_x, start_y, direction)
            if crosses >= 2:
                sum += 1
    return sum


if __name__ == '__main__':
    print(solve2(aoc.get_input_lines('day4.txt')))

File: day4.hs

-- | Helper function for cross
crossLine :: [a] -> b -> [(a, b)]
crossLine lst elem = map (,elem) lst

-- | Returns the cross product of two lists
cross :: [a] -> [b] -> [(a, b)]
cross xs = concatMap (crossLine xs)

-- | Safely access list at index, if index is negative or out of bounds, return Nothing
safeIndex :: [a] -> Int -> Maybe a
safeIndex lst i
    | i < 0             = Nothing
    | null (drop i lst) = Nothing
    | otherwise         = Just $ lst !! i

-- | word -> matrix of letters -> x -> y -> dx -> dy -> 1 if word exists, 0 otherwise.
-- x and y is where the word starts and dx dy is the direction the word goes in.
findWord :: String -> [String] -> Int -> Int -> Int -> Int -> Int
findWord "" _ _ _ _ _ = 1
findWord word letters x y dx dy = case safeIndex letters y of
    Nothing  -> 0
    Just row -> case safeIndex row x of
        Nothing     -> 0
        Just letter -> if letter == head word
                       then findWord (tail word) letters (x + dx) (y + dy) dx dy
                       else 0

-- | Given a matrix of letters and the x y start position of the word, return amount of instances
-- of the word "XMAS" in any of the eight orientations.
solveOne :: [String] -> Int -> Int -> Int
solveOne letters y x = sum $ map (uncurry $ findWord "XMAS" letters x y) $ cross [-1, 0, 1] [-1, 0, 1]

-- | Given a matrix of letters, return the amount of occurences of the word "XMAS"
solve :: [String] -> Int
solve inp = sum $ map (uncurry $ solveOne inp) $ cross [0..length inp - 1] [0..length (head inp) - 1]

main :: IO ()
main = do
    input <- readFile "day4.txt"

    let result = solve $ lines input

    print result